src/Analyser/NodeScopeResolver.php

```diff
@@ -5697,11 +5697,24 @@ private function processAssignVar(
 				} else {
 					$varForSetOffsetValue = $var->var;
 				}
-				$assignedPropertyExpr = new SetOffsetValueTypeExpr(
-					$varForSetOffsetValue,
-					$var->dim,
-					$assignedPropertyExpr,
-				);
+
+				if (
+					$var === $originalVar
+					&& $var->dim !== null
+					&& $scope->hasExpressionType($var)->yes()
+				) {
+					$assignedPropertyExpr = new SetExistingOffsetValueTypeExpr(
+						$varForSetOffsetValue,
+						$var->dim,
+						$assignedPropertyExpr,
+					);
+				} else {
+					$assignedPropertyExpr = new SetOffsetValueTypeExpr(
+						$varForSetOffsetValue,
+						$var->dim,
+						$assignedPropertyExpr,
+					);
+				}
```

Comment:
with this change, we make sure `PropertyAssignNode` contains a proper type.
we already had the right type in scope from a previous fix.